app/models/internal/scheduling/BaseSchedule.java

```diff
@@ -0,0 +1,136 @@
+     * @param builder Instance of <code>Builder</code>.
+     */
+    protected BaseSchedule(final Builder<?, ?> builder) {
+        _runAtAndAfter = builder._runAtAndAfter;
+        _runUntil = builder._runUntil;
+    }
+
+    public ZonedDateTime getRunAtAndAfter() {
+        return _runAtAndAfter;
+    }
+
+    public Optional<ZonedDateTime> getRunUntil() {
+        return _runUntil;
+    }
+
+    @Override
+    public Optional<ZonedDateTime> nextRun(final Optional<ZonedDateTime> lastRun) {
+        Optional<ZonedDateTime> result = unboundedNextRun(lastRun);
+        while (result.isPresent() && result.get().isBefore(_runAtAndAfter)) {
+            result = unboundedNextRun(result);
+        }
+        if (_runUntil.isPresent() && result.isPresent() && result.get().isAfter(_runUntil.get())) {
+            return Optional.empty();
+        }
+        return result;
+    }
+
+    /**
+     * Return the next time the schedule should run, much as {@link Schedule}{@code .nextRun} does.
+     * Will get restricted between {@code runAtAndAfter} and {@code runUntil} by {@code nextRun}.
+     *
+     * @param lastRun The last time the job was run.
+     * @return The next time to run the job.
+     */
+    public abstract Optional<ZonedDateTime> unboundedNextRun(Optional<ZonedDateTime> lastRun);
+
+    /**
+     * Builder implementation for {@link BaseSchedule} subclasses.
+     *
+     * @param <B> type of the builder
+     * @param <S> type of the object to be built
+     *
+     * @author Spencer Pearson (spencerpearson at dropbox dot com)
+     */
+    protected abstract static class Builder<B extends Builder<B, S>, S extends BaseSchedule> extends OvalBuilder<S> {
+        @NotNull
+        @ValidateWithMethod(methodName = "validateRunAtAndAfter", parameterType = ZonedDateTime.class)
+        protected ZonedDateTime _runAtAndAfter;
+        @NotNull
+        protected Optional<ZonedDateTime> _runUntil = Optional.empty();
+
+        /**
+         * Protected constructor for subclasses.
+         *
+         * @param targetConstructor The constructor for the concrete type to be created by this builder.
+         */
+        protected Builder(final Function<B, S> targetConstructor) {
+            super(targetConstructor);
+        }
+
+        /**
+         * Called by setters to always return appropriate subclass of
+         * <code>Builder</code>, even from setters of base class.
+         *
+         * @return instance with correct <code>Builder</code> class type.
+         */
+        protected abstract B self();
+
+        /**
+         * The earliest time at which the schedule should run.
+         *
+         * @param runAtAndAfter The time.
+         * @return This instance of {@code Builder}.
+         */
+        public B setRunAtAndAfter(final ZonedDateTime runAtAndAfter) {
+            _runAtAndAfter = runAtAndAfter;
+            return self();
+        }
+
+        /**
+         * The last time at which the schedule should ever run.
```

Comment:
We usually include whether the field is required or optional in the comment. Further, if optional we describe the default value as well. And either way describe any constraints (e.g. `Cannot be null`).

e.g.
https://github.com/ArpNetworking/metrics-aggregator-daemon/blob/master/src/main/java/com/arpnetworking/tsdcore/model/Quantity.java#L209